### ISSUE TITLE
Descriptive filenames and directory structure

### ISSUE THREAD

**Issue by timomeh** [feathersjs/generator-feathers#95]
As I [mentioned](https://twitter.com/timomeh/status/713387737047568385) on Twitter, all those index.js files are unpleasant and – in my opinion – a commonly used antipattern in Node.js. As an application grows, one will most likely end up with a big pile of index.js files.

Many index.js files will slow down every developer. The Tree-View in your Editor is confusing; too many Tabs are titled `index.js – src/se...`; you can't really use _"quick jump to a file"_ (e.g. cmd+T in Atom); …

Also it tends to an exaggerated directory structure. I think everyone has a project with this one directory which only contains a single file (index.js), and everytime you want to open that friggin' file you have to expand it in your tree view. In the scaffolded project there's `src/services/user/hooks`. It has only the index.js with the before- and after-hooks. I could come up with another file, maybe `user-hooks.js`, but after that I don't know what to put in this directory.

**tl; dr** a flatter file structure with reasonable filenames and directory-groups would be appropriate for the awesomeness of feathersjs.

To conclude with more specific ideas, a big change to a more widespread structure could be:

```
src/
  hooks/
    UserHooks.js
    MessageHooks.js
  middleware/
    [...]
  models/
    UserModel.js
    MessageModel.js
  services/
    UserService.js
    MessageService.js
```

Or it could be possible to group everything under `services/` (somewhat like it's now):

```
src/
  middleware/
    [...]
  services/
    users/
      UserHooks.js
      UserModel.js
      UserService.js
    messages/
      MessageHooks.js
      MessageModel.js
      MessageService.js
```

_Notice that with the second attempt the files are grouped by their service, but thus they need to be one level deeper in the directory structure if you compare it to the first example._

I used CamelCase filenames in my examples because it's more widespread, but I don't have anything against dash-case. CamelCase though can cause problems with sloppy programmers on non-case-sensitive filesystems (like the default JHFS+ in OS X).

So… yeah.

**Comment by thejones** [comment]
Just my thoughts for the discussion. Going with a dedicated models directory can assist with some of the Sequelize relationships issues that a few of us have experienced. https://github.com/feathersjs/feathers-docs/issues/51

```
src/
  models/
    UserModel.js
    ArticleModel.js
    index.js - Load/Sync models here
  services/
    user/
      hooks/ 
        UserMailer.js
        ValidateUser.js
      UserService.js
    articles/
      hooks/
        OwnResource.js
        CanPost.js
      ArticleService.js
  middleware/
    [...]
  
```

**Comment by ekryski** [comment]
I have to admit I'm pretty in favour of @timomeh's suggestion. Force of bad habit on my part I guess, but I'd agree that using `index.js` is kind of a bad anti-pattern.

@thejones I want to see what you and @DesignByOnyx are cooking up. I want to spend some time with this but I only concern with not having the model at the service level is how you do associations once things have become services. I've spiked remote services out with Mongoose but not with Sequelize so I'm also going to tinker with it to see if I can come up with a way to do many-to-many relationships that can scale across a server cluster.

Overall, lot's of :+1: :+1:'s over here. :smile:

**Comment by ekryski** [comment]
Somewhat related to #94.

**Comment by timomeh** [comment]
Another thing which is related to this is: how to deal best with versioned APIs. [The feathers-docs on versioning](http://docs.feathersjs.com/middleware/routing.html#versioning) describe to ways, as distinctive routes or as sub-apps, but the project structure is open to the developer (which is totally ok for documentation purposes).

I'll try out different structures in the next days and share my experiences with them.

**Comment by marshallswain** [comment]
The second layout definitely does more for the portability of hooks.  To split out a service that uses relational models, you'd only have to copy over the related model files into the service folder and fix the `require`s in the primary model.
```
src/
  middleware/
    [...]
  services/
    users/
      UserHooks.js
      UserModel.js
      UserService.js
    messages/
      MessageHooks.js
      MessageModel.js
      RelatedModel1.js
      RelatedModel2.js
      MessageService.js
```

**Comment by daffl** [comment]
As for removing the `index.js` files, I totally agree. They do contain some important code though but things like `services/index.js` can be easily moved to just `services.js` and the path to import them would still look the same.

In the last 8+ years of doing JavaScript I had a discussion or two about file structures :wink: so here are some more of my thoughts:

- Things should be structured on a per service (which is the main entity in Feathers) level. At Bitovi for UI components we call it a [modlet](https://donejs.com/Features.html#section=section_Modlets). Separate top level folders (`/hooks/`, `/services/` etc.) make it really hard to navigate larger apps.
- I'm not sure if having multiple files with the same name is actually a problem or just a personal preference. Is there any IDE or text editor where finding `messages/MessagesService.js` is easier than searching for `messages/service.js`?
- No capitalized file names. It's really no fun dealing with in different operating systems. I just recently ran into a problem of working on a case insensitive file system but the CI server was on Linux so it worked locally but not on CI and there was no easy way to change the casing via Git (because it didn't pick it up as a change locally). Since everybody seems to lean in that direction, I'd be totally ok with more descriptive naming of files with something like `messages/messages-service.js` or `messages/messages.service.js`.

**Comment by timomeh** [comment]
@daffl 
> I'm not sure if having multiple files with the same name is actually a problem or just a personal preference. Is there any IDE or text editor where finding messages/MessagesService.js is easier than searching for messages/service.js?

I tested this in Atom, and it's easier to find `messages/MessagesService.js` than `messages/service.js`. Atoms Fuzzy File Finder prioritizes filenames. Directories are listed after all fuzzy-found filenames, and even they are listed by results in the filename. Of course you can find the file easily if you type `messages/service.js`, but the whole point of this is to find files quick with a few keystrokes. Sublime Text behaves nearly the same.

I'm a big fan of optimizing HCI. For example, if you have a tree-view with multiple identical filenames, your eyes have to scan both filename and directory-name to find the correct file (which results in unnecessary eye-movement, distraction, ...). Although these are really small things, it does a lot to feel more comfy while developing.

**Comment by daffl** [comment]
That's good enough for me :smile: As long as we stick with lowercase filenames and decide between `user-service.js` and `user.service.js` or something similar the next version will have a better file structure.

**Comment by timomeh** [comment]
Sounds good to me 😊

**Comment by marshallswain** [comment]
I would appreciate dropping index.js everywhere.  Then I could stop using Atom's [foldername-tabs plugin](https://atom.io/packages/foldername-tabs).

**Comment by alnour-altegani** [comment]
when will you change the directory structure ? like
```bash
/hooks
/middlewares
/services
   /messages/service.js
   /messages/model.js
   /messages/hooks.js
```

**Comment by ekryski** [comment]
@alnour-altegani it might be a bit still but we have started to work on the new CLI which will generate the new structure. We're not going to modify the existing generator much.

**Comment by jay-at-avido** [comment]
+1000 for issue!

Personally this has been one of most annoying this with FeathersJS workflow so far.
I want to second what David said when he cites that grouping files by feature is usually a much better choice since it allows for better modularity and overall project scalability. 

Obviously this issue has been discussed many times over before in the context of other web frameworks, so maybe the Feathers community can benefit from that previous knowledge. 
Here's a link that comes to mind about the same discussion but in the context of RubyOnRails and the Trailblazer frameworks: 
https://news.ycombinator.com/item?id=8671590

And because one picture is worth more than a thousand words:
[index.js wat?](https://dl.dropboxusercontent.com/u/1631393/test/index_wat.png)

**Comment by marshallswain** [comment]
I totally understand any feelings you might have about the overabundance of `index.js` files.  It would be a big help just to rename each index.js to match its folder name, instead of using the node convention.

**Comment by jay-at-avido** [comment]
I'll also add this a suggestion regarding the structure for hooks: 

Personally I tend to favor solutions that try and keep all (or at least most of) the middleware involved on any given request/response pipeline at view on a **single file**.

Because of this, I'd much rather have this kind of structure for the hooks functionality, be it a top-of-the-file solution like Rails `before_filter:`/`after_filter:`  or something akin to Flask's (View Decorators)[http://flask.pocoo.org/docs/0.11/patterns/viewdecorators/]

Notice that only the reference is kept on the same file, actual function implementation is kept on  separated files so there's no clutter, but if you are good at naming your functions, than that's all you'll need to have a complete picture of what middleware is acting upon your data

Do anyone else feel that they have to always keep both the hooks file and the service file at close sight at the same time, or is just me?
